### ISSUE TITLE
Loggers don't work properly when master=False

### ISSUE THREAD

**Issue by diranged** [unbit/uwsgi#894]
It seems that if you add `master=False` to your uWSGI settings (in both 1.4.x and 2.0.10), adding a logger (ie, `logger=syslog:myapp,local0`) simply fails to log anything at all.

**Comment by unbit** [comment]
You cannot use advanced loggers without a master as their hook functions are run into it (workers simply pass log strings via a socketpair() )

**Comment by robotclarke** [comment]
This is also hurting me as I'm trying to run uWSGI as one process (with multiple Python threads) inside a Docker container.  Do you have any workarounds to change the logger/logformat with `master = false`?

**Comment by unbit** [comment]
there is no workaround, pluggable loggers are all written for the master. Why you want to remove the master ?

**Comment by robotclarke** [comment]
It's simpler for me to manage one process per container as we're using Amazon ECS to dynamically provision and deploy new containers - if the worker dies then the container dies and ECS restarts it.

With the master process we get another layer of complexity where if the worker dies, the master will attempt to restart it, and we have less visibility over that happening (in fact one reason we switched away from gunicorn to uWSGI is because our workers kept dying and getting restarted inside the container for some reason, killing our performance).  I notice you've done a similar thing with your [uwsgi-docker](https://github.com/unbit/uwsgi-docker) project using emperor but it looks like you're still running a master/multiple workers inside the vassals, which I'd rather not do.

As an aside I'm having a lot of problems with the logging as we're producing JSON logs (hence why I need the log-encoder) but the log-encoder only processes the lines right before they go to stdout, so I can't get any of the logformat variables into separate JSON keys.  Should I be writing my own plugin to handle the logging, and if I do will it make the `master = false` case work as well?  Thanks for the help :smile:

**Comment by unbit** [comment]
Yes in uwsgi-docker there is integration with Emperor and vassals (and masters obviously). Workers tend to dies too often to restart the container every time it happens :) Btw, the only solution i can think of (expecially because when you send logs to external services you absolutely need a dedicated thread/process otherwise logging could became a bottleneck) is allowing the spawn of the logger_thread in the process (if there is only one worker)

**Comment by robotclarke** [comment]
That would be good, it seems when `master = false` all the worker threads are responsible for their own logging.  It causes threading problems and mangled logs for me as well as the bottlenecks you mentioned :grinning: 

```
10.0.2.2 - GET /static/1.png HTTP/1.1 404 9ms
10.0.2.210.0.2.2 - GET /static/3.png HTTP/1.1 404 8ms
 - GET /static/2.png HTTP/1.1 404 4ms
10.0.2.2 - GET /static/4.png HTTP/1.1 404 4ms
```

Allowing the `logger_thread` to start up without the master would be nice (looks like it's spawning inside a `if (uswgi.log_master)` block atm); I don't know what knock on effects that would have though for the rest of the codebase.

**Comment by unbit** [comment]
This commit https://github.com/unbit/uwsgi/commit/20a0be1897d8688bae94bf4f52be6cc259ce8759 introduces --worker-logger, --worker-logger-req, --worker-log-encoder and worker-log-req-encoder options. They can be used only only in single process mode. It is not clear if some corner case can pop-up. Waiting to see your report

**Comment by robotclarke** [comment]
Finally got round to testing this and it works great, thanks for all the help!  No more threading issues, and all the log-encoders I've tested appear to work.  Hit a bit of a roadblock with #1026, but that's a different matter.  Also I think I'd prefer it if the original --logger option worked as expected here instead of a separate argument, but I'm not complaining!  (I guess I can't see why anyone would want the previous behaviour of all threads doing their own logging.)

For anyone that finds this later: if you want logging to stdout but with a log-encoder, you can set as many --worker-log-encoders as you want but you need to set the --worker-logger to enable it, otherwise nothing will get logged.  Try `--worker-logger stdout fd:1` to replicate the normal logger's default functionality.
